internal/controller/quota_tracker.go

```diff
@@ -0,0 +1,149 @@
+
+// A tracker of allocated quota, mapped by namespace
+type QuotaTracker struct {
+	// state of quotas, used, and allocated amounts
+	state map[string]*QuotaState
+
+	// used amounts by dispatched AppWrappers partially unaccounted by the ResourceQuota,
+	// as some pods may have not passed the ResourceQuota admission controller
+	unAdmittedWeightsMap map[string]*WeightsPair
+}
+
+// Create a new QuotaTracker
+func NewQuotaTracker() *QuotaTracker {
+	return &QuotaTracker{
+		state:                map[string]*QuotaState{},
+		unAdmittedWeightsMap: map[string]*WeightsPair{},
+	}
+}
+
+// State includes total quota, used quota, and currently allocated quota
+type QuotaState struct {
+	// quota enforced in the ResourceQuota object
+	quota *WeightsPair
+	// used amount in the status of the ResourceQuota object
+	used *WeightsPair
+	// allocated amount by dispatched AppWrappers in the current dispatching cycle
+	allocated *WeightsPair
+}
+
+// Create a QuotaState from a ResourceQuota object
+func NewQuotaStateFromResourceQuota(resourceQuota *v1.ResourceQuota) *QuotaState {
+	quotaWeights, usedWeights := getQuotaAndUsedWeightsPairsForResourceQuota(resourceQuota)
+	return &QuotaState{
+		quota:     quotaWeights,
+		used:      usedWeights,
+		allocated: NewWeightsPair(Weights{}, Weights{}),
+	}
+}
+
+// Account for all in-flight AppWrappers with their resource demand not yet reflected in
+// the Used status of any ResourceQuota object in their corresponding namespace
+func (tracker *QuotaTracker) Init(weightsPairMap map[string]*WeightsPair) {
+	tracker.unAdmittedWeightsMap = weightsPairMap
+}
+
+// Check if the resource demand of an AppWrapper satisfies a ResourceQuota,
+// without changing the current quota allocation, returning resource names with insufficient quota
+func (tracker *QuotaTracker) Satisfies(appWrapperAskWeights *WeightsPair, resourceQuota *v1.ResourceQuota) (bool, []v1.ResourceName) {
+	namespace := resourceQuota.GetNamespace()
+	var quotaState *QuotaState
+	var exists bool
+	if quotaState, exists = tracker.state[namespace]; !exists {
+		quotaState = NewQuotaStateFromResourceQuota(resourceQuota)
+		tracker.state[namespace] = quotaState
+	}
+	// check if both appwrapper requests and limits fit available resource quota
+	quotaWeights := quotaState.quota.Clone()
+	quotaWeights.Sub(quotaState.used)
+	quotaWeights.Sub(quotaState.allocated)
+	var unAdmittedWeights *WeightsPair
+	if unAdmittedWeights, exists = tracker.unAdmittedWeightsMap[namespace]; exists {
+		quotaWeights.Sub(unAdmittedWeights)
+	}
+	quotaFits, insufficientResources := appWrapperAskWeights.Fits(quotaWeights)
+
+	mcadLog.Info("QuotaTracker.Satisfies():", "namespace", namespace,
+		"QuotaWeights", quotaState.quota, "UsedWeights", quotaState.used,
+		"AllocatedWeights", quotaState.allocated, "unAdmittedWeights", unAdmittedWeights,
+		"AvailableWeights", quotaWeights, "appWrapperAskWeights", appWrapperAskWeights,
+		"quotaFits", quotaFits)
```

Comment:
Please disable this log line by default as it is going to make logging very verbose.

Reply:
done